### PR TITLE
Close #147: postmarketos-base: added udev rule and script for firmwar…

### LINKED ISSUE #147: Package udev config for firmware
@PabloCastellano got the firmware loading working in #143 but I think that the `firmware.sh` file and the udev rule can be moved to a seperate package (or postmarketos-base) and modify `firmware.sh` to also search files in `/lib/firmware/postmarketos`.

This system should work for drivers following the kernel way of loading firmware and will become more important for mainlined devices.

### PR DESCRIPTION
**Pull request by PabloCastellano**
…e loading

### CODE REVIEW

File: aports/main/postmarketos-base/50-firmware.rules [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# do not edit this file, it will be overwritten on update
+SUBSYSTEM=="firmware", ACTION=="add", RUN+="/usr/bin/firmware.sh"
```
Comment (ollieparanoid) [inline aports/main/postmarketos-base/50-firmware.rules]: From [FHS](http://www.pathname.com/fhs/pub/fhs-2.3.html#AEN1389):
> `/usr/lib` includes object files, libraries, and internal binaries that are not intended to be executed directly by users or shell scripts.

...so I would put the `firmware.sh` script in `/usr/lib/`. You have called the logfile `firmwarelog`, which seems to catch the purpose a bit better, so I recommend naming the script `/usr/lib/firmwareload.sh`.

File: aports/main/postmarketos-base/firmware.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#!/bin/sh
+# udev dynamic firmware loading script
+########################################################################
+
+FIRMWARE_DIRS="/lib/firmware /lib/firmware/postmarketos"
+
+exec >>/root/firmwareload.log 2>&1
```
Comment (ollieparanoid) [inline aports/main/postmarketos-base/firmware.sh]: I think `/var/log/firmwareload.log` would be a better place for the log (if we have used `/root` somewhere else, please tell me).

File: aports/main/postmarketos-base/firmware.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+if [ ! -e /sys/$DEVPATH/loading ]; then
+    echo "firmware loader misses sysfs directory"
+    exit 0
```
Comment (ollieparanoid) [inline aports/main/postmarketos-base/firmware.sh]: Is that intended, or should this be `exit 1`?

File: aports/main/postmarketos-base/firmware.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    [ -e "$DIR/$FIRMWARE" ] || continue
+    echo "loading $DIR/$FIRMWARE"
+    echo 1 > /sys/$DEVPATH/loading
+    cat "$DIR/$FIRMWARE" > /sys/$DEVPATH/data
+    echo 0 > /sys/$DEVPATH/loading
+    exit
+done
+
+echo -1 > /sys/$DEVPATH/loading
+echo "Cannot find  firmware file '$FIRMWARE'"
```
Comment (ollieparanoid) [inline aports/main/postmarketos-base/firmware.sh]: Two spaces between `find  firmware`, also this log message is started with a capital letter, the others are not. Consistency within the same log file would be nice :)

File: aports/main/postmarketos-base/firmware.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
+#!/bin/sh
+# udev dynamic firmware loading script
+########################################################################
```
Comment (ollieparanoid) [inline aports/main/postmarketos-base/firmware.sh]: We don't have such lines in other shell scripts I think, so for consistency I would remove it.

**Comment by PabloCastellano** [comment]
Thanks for the review @ollieparanoid. As always, your comments make perfectly sense :smiley: 

I've also done some changes to make shellcheck happier.

Maybe we could also copy this part from yuvadm's [firmware.sh](https://github.com/yuvadm/pmbootstrap/blob/device-mozilla-flame/aports/device/device-mozilla-flame/firmware.sh#L24-L29). But I prefer to have it already merged and then make the additional improvements

**Comment by ollieparanoid** [comment]
*Just labeled, because it became clear in the chat that it isn't easy to see which PRs are ready to merge and which are not. Please remove the label when you think it's ready!*

**Comment by yuvadm** [comment]
@PabloCastellano are we stalled on this? I'd be happy to amend the PR per @ollieparanoid's comments, would love to see this merged.

**Comment by ollieparanoid** [comment]
@yuvadm: the whole point of making the PR branches directly in the `pmbootstrap` repo is, that others can continue the work - so I'm pretty sure it is okay with him if you did just that 👍

**Comment by yuvadm** [comment]
@ollieparanoid updated per your review

**Comment by PabloCastellano** [comment]
Oops, turns out that I did the changes but pushed them to [my repo](https://github.com/PabloCastellano/pmbootstrap/commits/firmware-dynamic-loading) instead of pmbootstrap

File: aports/main/postmarketos-base/APKBUILD [hunk 6]
```diff
@@ -10,12 +10,19 @@ tmux htop nano postmarketos-mkinitfs eudev sudo"
...
-source="sudoers $install"
+source="sudoers firmware.sh 50-firmware.rules $install"
 options="!check"
 
 package() {
 	install -D -m644 "$srcdir"/sudoers \
 		"$pkgdir"/etc/sudoers.d/postmarketos
+	install -D -m644 "$srcdir"/50-firmware.rules \
+		"$pkgdir"/etc/udev/rules.d/50-firmware.rules
+	install -D -m755 "$srcdir"/firmware.sh \
```
Comment (PabloCastellano) [inline aports/main/postmarketos-base/APKBUILD]: To make things easier I'd rename firmware.sh to firmwareload.sh and then just copy

**Comment by PabloCastellano** [comment]
I've pushed my branch again as [firmware-dynamic-loading-2](https://github.com/postmarketOS/pmbootstrap/tree/firmware-dynamic-loading-2). Feel free to merge it or discard.

**Comment by ollieparanoid** [comment]
I have compared both PRs, and I'm going with #427 instead, because it resolved the merge conflict already, renamed the script to `firmwareload.sh` and added automatic checking for that script in `static_code_analysis.sh`.

Sorry for the duplicate effort @yuvadm. But at least it can finally get merged!

**Comment by yuvadm** [comment]
@ollieparanoid No problem, just happy we got this merged :) thanks for the work @PabloCastellano!

**Pull request closed without merging**